Feature scraper

ElenaKusevska opened the pull request.

ElenaKusevska committed:
Traverse categories

ElenaKusevska committed:
Get article html, add main categories to homepage

ElenaKusevska committed:
get article title, intro, and methods

ElenaKusevska committed:
Get the title, text and image for each step in the methods

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+        # where each subcategory was referenced from
+
+        main_wiki_url = "https://www.wikihow.com"
+        sitemap_url = "https://www.wikihow.com/Special:Sitemap"
```
No need to hard-code wikiHow EN. I believe this page is available at same path on all websites

ElenaKusevska replied:
@rgaudin, you're right, I just realized it's just self.url

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+
+            methods.append(method)
+
+        sys.exit()
```
`sys` is not imported. Again, you should integrate `flake8` in your text editor or run it periodically as you have several undeclared variables in this code

ElenaKusevska replied:
@rgaudin, I don't intend on using sys. This is just a draft. I'm just using this line to mark where I'm working on. I'll delete it.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+        category_divs = soup.find_all("div", {"class": "cat_list"})
+        self.main_categories = []
+        for div in category_divs:
+            h3 = div.find("h3")
```
You may want to use a different selector to directly retrieve the `h3`

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+            h3 = div.find("h3")
+            cat_url = main_wiki_url + h3.find("a")["href"]
+            print("0 category " + cat_url, flush=True)
+            current_category = [h3.text, self.output_dir + "/C" + h3.find("a")["href"]]
```
This looks like you are trying to replicate the kiwix-serve path here.

When you add path `home` to a zim (`.add_content_for(path="home", content="hello")`), it is reachable by kiwix-serve at `http://{host}:{port}/{zim_name}/C/home` or even without the `C/` if you are using the latest (unreleased I believe) kiwix-serve.
You **should not** care about this prefix as this is a kiwix-serve only prefix that is used to seggregate zim files (kiwix-serve can serve multiple zim files).

ElenaKusevska replied:
@rgaudin This isn't for the path when creating the zim, it's for the link. I realized that in-zim links don't work unless the entire path is written with kiwix-serve

ElenaKusevska replied:
Nevermind, I figured it out :)

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+            logger.critical(f"Unable to retrieve homepage at {cat_url}: {exc}")
+            logger.exception(exc)
+
+        soup = bs4.BeautifulSoup(response.text, "html.parser")
```
Already mentioned that we want to use `lxml` parser everywhere. `html.parser` is too slow.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+        article_divs = soup.find_all("div", {"class": "responsive_thumb"})
+        for article_div in article_divs:
+            article_url = article_div.find("a")["href"]
+            if not article_url in self.articles_list:
```
convention is to use `if xxx not in yyy` not `if not xxx in yyy`.

Also, the is a poorly chosen name for a `dict`: `_list` implies a type which is different.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+        #     )
+        # )
+
+    def walk_subcategories(self, main_wiki_url, cat_url, recursion_depth):
```
Why would we need to pass the static `main_wiki_url` ?

ElenaKusevska committed:
Edit render article

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+        for article_div in article_divs:
+            article_url = article_div.find("a")["href"]
+            if not article_url in self.articles_list:
+                self.articles_list[article_url] = [str(recursion_depth) + " " + cat_url]
```
what's the purpose of this ? Debug print only?

ElenaKusevska replied:
Yes, I used this to see how deep the recursion goes, and also how many places an article is accessed from. I can delete these lines when I'm done.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -76,12 +76,161 @@ def add_assets(self):
+        # Get all the articles by walking through all the main categories
+        # and the sub-categories
+
+        self.articles_list = {}  # keep track of each article processed and
```
Don't create scraper instance properties inside methods if those are to be used anyway. This should be moved to __init__

ElenaKusevska replied:
@rgaudin Let's Keep them for now. We can remove them later on.

rgaudin reviewed (commented):
A few hints on your WIP

ElenaKusevska committed:
pass article content to article template

ElenaKusevska committed:
Add pages to list caegories and articles per category

ElenaKusevska committed:
remove hard coded main wiki url

ElenaKusevska committed:
change to lxml parser

Commit:
reorganised inputs handling

- reusing separate Conf approach from sotoki (less boilerplate in scraper)
- added inputs sanitization, handling and default values
- fixed languages mapping: using language codes as keys (not tld)
- TEMP: removed walker code. will add back iteratively

Commit:
Basic scraper producing homepage

New *destructive* approach for scraping:
- Use a base similar to wikiHow's design
- Include all of source's stylesheets and resources
- extract the main content part of pages we want
- remove the extra elements we don't want
- rewrite result and add to template

Reusing parts of sotoki: executor, imager, global context
- Rewriter to fix links and images from extracted source
- Imager to process images during rewrite
- Executor to work with Imager on multiple threads

Also:
- Now fetching various metadata from source website
- Properly using/setting all Zim metadata, including illustrations and favicon
- CSS parser that looks for url(), downloads and add them to Zim and replace path
- Custom CSS to override some rules to our liking (very limited)
- --category param to specify one or several categories to scrape to limit processing
to a subset. Defaults to finding all root categories
- Recursively looks for sub categories
- Better log messages
- Basic libzim error handling
- Initial images support with S3 (from sotoki)
- moved all assets to `assets/` instead of `assets/static/`

Commit:
Added category pages support

- Added category-specific CSS
- walks categories and sub categories
- fetch list of articles from category
- supports pagination in category pages
- add category pages to zim
- Supports webp images from wikiHow
Category images are referenced using <img /> without src attr:
- a data-src one with a WebP url
- a data-nowebp one with a JPG url
- a following <img /> with a JPG src wrapped in a <noscript />
- added number of images to stats
- Hiding expanding-icon for sub-sub cats as it requires JS on source website

Commit:
Added Article scraping support

- Articles added to Zim
- Article template contains most info
/!\ some parts still buggy
